src/stamps.ts

```diff
@@ -1,28 +1,235 @@
+/**
+ * Calculate usage of a given postage stamp
+ *
+ * @param stamp Postage stamp which usage should be determined
+ */
+export function getUsage({ utilization, depth, bucketDepth }: DebugPostageBatch): number {
+  return utilization / Math.pow(2, depth - bucketDepth)
 }
 
 /**
- * Return true if the proxy should replace the postage stamp header
+ * Filter the stamps and only return those that are usable, have correct amount, depth, are not close to beying maxUsage or close to expire
+ *
+ * @param stamps Postage stamps to be filtered
+ * @param depth Postage stamps depth
+ * @param amount Postage stamps amount
+ * @param maxUsage Maximal usage of the stamp to be usable by the proxy
+ * @param minTTL Minimal TTL of the stamp to be usable by the proxy
+ *
+ * @returns Filtered stamps soltered by usage
+ */
+export function filterUsableStamps(
+  stamps: DebugPostageBatch[],
+  depth: number,
+  amount: string,
+  maxUsage: number,
+  minTTL: number,
+): DebugPostageBatch[] {
+  const usableStamps = stamps
+    // filter to get stamps that have the right depth, amount and are not fully used or expired
+    .filter(s => s.usable && s.depth === depth && s.amount === amount && getUsage(s) < maxUsage && s.batchTTL > minTTL)
+    // sort the stamps by usage
+    .sort((a, b) => (getUsage(a) < getUsage(b) ? 1 : -1))
+
+  // return the all usable stamp sorted by usage
+  return usableStamps
+}
+
+/**
+ * Buy new postage stamp and wait until it is usable
+ *
+ * @param depth Postage stamps depth
+ * @param amount Postage stamps amount
+ * @param beeDebug Connection to debug endpoint for checking/buying stamps
+ * @param pollingFrequency (optional) How often should the stamp be checked in ms, default 1000
+ *
+ * @returns Newly bought postage stamp
  */
-export function shouldReplaceStamp({ POSTAGE_STAMP }: PostageStamps = {}): boolean {
-  return Boolean(POSTAGE_STAMP)
+export async function buyNewStamp(
+  depth: number,
+  amount: string,
+  beeDebug: BeeDebug,
+  pollingFrequency = 1_000,
+): Promise<DebugPostageBatch> {
+  return new Promise(async (resolve, reject) =>
+    beeDebug
+      .createPostageBatch(amount, depth)
+      .then(async (batchId: BatchId) => {
+        waitUntilStampUsable(batchId, beeDebug, pollingFrequency)
+          .then(resolve)
+          .catch(e => {
+            logger.error('error: failed to wait until postage stamp is usable', e)
+            reject(e)
+          })
+      })
+      .catch(e => {
+        logger.error('error: failed to buy postage stamp', e)
+        reject(e)
+      }),
+  )
+}
+
+export class StampsManager {
+  private stamp?: string
+  private usableStamps?: DebugPostageBatch[]
+  public readonly enabled: boolean
+  private interval?: ReturnType<typeof setInterval>
+
+  constructor(
+    {
+      POSTAGE_STAMP,
+      POSTAGE_DEPTH,
+      POSTAGE_AMOUNT,
+      POSTAGE_USAGE_THRESHOLD,
+      POSTAGE_USAGE_MAX,
+      POSTAGE_TTL_MIN,
+      POSTAGE_REFRESH_PERIOD,
```

Comment:
I would not use upper camelcase for these. While yes the values are passed from env. variables for which I would also use this style, in this context they are just simple parameters that somebody passes in somehow. This is also a bit confusing as this style is used for constants which these are not.